### DISCUSSION TITLE
How use msw in browser mode

### DISCUSSION THREAD

**Discussion by njzydark** [vitest-dev/vitest#6231] [Q&A] [answered]
I use `msw/browser` to setup worker, but when i await `worker.start`, it stuck

```ts
import { setupWorker } from 'msw/browser';
import { afterEach, beforeAll } from 'vitest';

const server = setupWorker([]);

beforeAll(async () => {
  await server.start();
});

afterEach(() => {
  server.resetHandlers();
});
```

**Comment by edwincarlflores** [comment]
I've been trying to create a utility function similar to how I do it with msw's setup server:
```ts
export function startMockServer(handlers: HttpHandler[]) {
  const server = setupServer(...handlers)
  beforeAll(() => server.listen({ onUnhandledRequest: 'error' }))
  afterEach(() => server.resetHandlers())
  afterAll(() => server.close())

  return server
}
```

But it seems that it is not working for msw's browser worker. What worked for me is this:

```ts
// browser.ts
import { setupWorker } from 'msw/browser'

export const worker = setupWorker()
```

Then on your test file:

```ts
import { worker } from '../../mocks/browser'
import { handlers } from '../../mocks/handlers'

describe('test', async () => {
  await worker.start()
  worker.use(...handlers)

  test(...)
})
```

**Reply by ysgk** [reply]
It seems that `worker.start()` should be called in each test file instead of the setup file. Although writing it in every test file is a bit tedious.

**Reply by charlieforward9** [reply]
@ysgk is this still the case? I am having insane headaches trying to get browser mode to work... see https://github.com/mswjs/msw/issues/2398

**Reply by ysgk** [reply]
@charlieforward9 
No, I’m not experiencing it anymore. In my case, the issue was fixed in version 2.1.4 by https://github.com/vitest-dev/vitest/pull/6772

**Comment by kettanaito** [comment]
I've added the [Vitest Browser Mode recipe](https://mswjs.io/docs/recipes/vitest-browser-mode) to the MSW docs. Please follow it to integrate MSW with Vitest Browser Mode. Thanks!

**Reply by charlieforward9** [reply]
If theres any way this issue can be resolved relating to `vitest` and `msw` using the browser config, would be incredible

https://github.com/vitest-dev/vitest/issues/7290

**Comment by kettanaito** [comment, marked as answer]
Hi. Please see the [Vitest Browser Mode](https://mswjs.io/docs/recipes/vitest-browser-mode) recipe. It describes how to integrate MSW in Vitest Browser Mode.

**Reply by charlieforward9** [reply]
Hey @kettanaito thanks for making this available.

Im not sure if this message is in response to [mine](https://github.com/vitest-dev/vitest/discussions/6231#discussioncomment-11938208), but were having a lot of trouble mocking other modules. 

I dont see any mention of that within the vitest browser docs, I would like to get around this, but want to make sure this recipe either solves it or if not, there is an idea for moving forward in the short term.

**Reply by kettanaito** [reply]
That one is a bug and it will be addressed as soon as someone has time. Hopefully soon! You are also welcome to dive in and do some investigation as to what is causing that behavior. Logging out the worker and its handlers it a good place to start. [Debugging runbook](https://mswjs.io/docs/runbook) also applies.
